plugin/path_checkouts.go

```diff
@@ -5,11 +5,349 @@ import (
+			continue
+		}
+		respData := map[string]interface{}{
+			"service_account_name": serviceAccountName,
+			"password":             password,
+		}
+		internalData := map[string]interface{}{
+			"service_account_name": serviceAccountName,
+		}
+		resp := b.Backend.Secret(secretAccessKeyType).Response(respData, internalData)
+		resp.Secret.Renewable = true
+		resp.Secret.TTL = ttl
+		resp.Secret.MaxTTL = set.MaxTTL
+		return resp, nil
+	}
+
+	// If we arrived here, it's because we never had a hit for a service account that was available.
+	resp, err := logical.RespondWithStatusCode(&logical.Response{
+		Warnings: []string{"No service accounts available for check-out, please try again later."},
+	}, req, 429)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (b *backend) checkOut(ctx context.Context, storage logical.Storage, serviceAccountName string, newCheckOut *CheckOut) (string, error) {
+	lock := locksutil.LockForKey(b.checkOutLocks, serviceAccountName)
+	lock.Lock()
+	defer lock.Unlock()
+
+	// We've found a service account that we can check out. Let's do it and respond.
+	if err := b.checkOutHandler.CheckOut(ctx, storage, serviceAccountName, newCheckOut); err != nil {
+		return "", err
+	}
+	password, err := retrievePassword(ctx, storage, serviceAccountName)
+	if err != nil {
+		return "", err
+	}
+	return password, nil
+}
+
+func (b *backend) secretAccessKeys() *framework.Secret {
+	return &framework.Secret{
+		Type: secretAccessKeyType,
+		Fields: map[string]*framework.FieldSchema{
+			"service_account_name": {
+				Type:        framework.TypeString,
+				Description: "Access Key",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Secret Key",
+			},
+		},
+		Renew:  b.renewCheckOut,
+		Revoke: b.endCheckOut,
+	}
+}
+
+func (b *backend) renewCheckOut(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	serviceAccountName := req.Secret.InternalData["service_account_name"].(string)
+
+	lock := locksutil.LockForKey(b.checkOutLocks, serviceAccountName)
+	lock.Lock()
+	defer lock.Unlock()
+
+	checkOut, err := b.checkOutHandler.Status(ctx, req.Storage, serviceAccountName)
+	if err != nil {
+		return nil, err
+	}
+	if checkOut.IsAvailable {
+		// It's possible that this renewal could be attempted after a check-in occurred either by this entity or by
+		// another user with access to the "manage check-ins" endpoint that forcibly checked it back in.
+		return logical.ErrorResponse(fmt.Sprintf("%s is already checked in, please call check-out to regain it", serviceAccountName)), nil
+	}
+	return &logical.Response{Secret: req.Secret}, nil
+}
+
+func (b *backend) endCheckOut(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	serviceAccountName := req.Secret.InternalData["service_account_name"].(string)
+
+	lock := locksutil.LockForKey(b.checkOutLocks, serviceAccountName)
+	lock.Lock()
```

Comment:
What is this lock protecting?

Reply:
It's keeping anyone else from attempting to perform a check-out during the renewal process. Contending processes would be attempting to get a write lock to perform a check-out.

Reply:
This isn't existent anymore due to set-level locking, but FWIW, at the set level I replaced it with a read-lock and it didn't trigger any problems in my race-detection test, so I think we're good to go.